bitbucket_server.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	fluxapi "github.com/fluxcd/flux/pkg/api"
+	fluxapi_v9 "github.com/fluxcd/flux/pkg/api/v9"
+	"github.com/google/go-github/v28/github"
+)
+
+const BitbucketServer = "BitbucketServer"
+
+func init() {
+	Sources[BitbucketServer] = handleBitbucketServerPush
+}
+
+func handleBitbucketServerPush(s fluxapi.Server, key []byte, w http.ResponseWriter, r *http.Request) {
+	// See incomplete docs: https://confluence.atlassian.com/bitbucketserver/event-payload-938025882.html
+
+	body, err := github.ValidatePayload(r, key)
+	if err != nil {
+		http.Error(w, "The signature header is invalid.", http.StatusUnauthorized)
+		log(GitHub, "invalid signature:", err.Error())
+		return
+	}
+	if eventKey := r.Header.Get("X-Event-Key"); eventKey != "repo:refs_changed" {
+		http.Error(w, "Unexpected or missing header X-Event-Key", http.StatusBadRequest)
+		log(BitbucketServer, "unexpected X-Event-Key header:", eventKey)
+		return
+	}
+
+	event := struct {
+		Repository struct {
+			Links struct {
+				Clone []struct {
+					Href string `json:"href"`
+					Name string `json:"name"`
+				}
+			}
+		}
+		Changes []struct {
+			RefID string `json:"refid"`
+		}
+	}{}
+	if err := json.Unmarshal(body, &event); err != nil {
+		http.Error(w, "Unable to JSON decode payload", http.StatusBadRequest)
+		log(BitbucketServer, "unable to decode payload:", err.Error())
+		return
+	}
+
+	repo := ""
+	for _, link := range event.Repository.Links.Clone {
+		if link.Name == "ssh" {
+			repo = strings.TrimPrefix(link.Href, "ssh://")
+			break
+		}
+	}
+	if repo == "" {
+		http.Error(w, "Missing repository SSH clone link", http.StatusBadRequest)
+		log(BitbucketServer, "missing repository SSH clone link")
+		return
+	}
+
+	const branchPrefix = "refs/heads/"
+	ctx, cancel := context.WithTimeout(r.Context(), timeout)
+	defer cancel()
+	for _, v := range event.Changes {
+		if !strings.HasPrefix(v.RefID, branchPrefix) {
```

Comment:
We could check for `type == "BRANCH"` here instead.

Reply:
Done.